File: gastromanager-frontend/src/app/features/management/application/mappers/cash-register-session.mapper.ts
import {
  CashRegisterSession,
  PaymentMethodSummary,
  SessionSummary
} from '../../domain/models/cash-register-session.interface';
import {
  CashRegisterSessionResponseDto,
  PaymentMethodSummaryDto,
  SessionSummaryDataDto
} from '../../infrastructure/api/dtos/cash-register-session.dto';

// Mapper para CashRegisterSession - Actualizado para manejar nueva estructura
export function mapToCashRegisterSession(dto: CashRegisterSessionResponseDto): CashRegisterSession {
  return {
    uuid: dto.uuid,
    cashRegisterUuid: dto.cashRegisterUuid,
    cashRegisterName: dto.cashRegisterName,
    cashRegisterLocation: dto.cashRegisterLocation,
    initialAmount: dto.openingAmount || dto.initialAmount || 0,
    finalAmount: dto.closingAmount || dto.finalAmount,
    openedBy: dto.createdBy || dto.openedBy || 'unknown',
    closedBy: dto.closedBy,
    openedAt: dto.openingTime || dto.openedAt || dto.createdDate,
    closedAt: dto.closingTime || dto.closedAt,
    notes: dto.notes,
    status: dto.status as 'ACTIVE' | 'CLOSED'
  };
}

// Mapper para PaymentMethodSummary
function mapToPaymentMethodSummary(dto: PaymentMethodSummaryDto): PaymentMethodSummary {
  return {
    paymentMethodName: dto.paymentMethodName,
    paymentMethodDescription: dto.paymentMethodDescription,
    totalAmount: dto.totalAmount,
    transactionCount: dto.transactionCount
  };
}

// Mapper para SessionSummary
export function mapToSessionSummary(dto: SessionSummaryDataDto): SessionSummary {
  return {
    sessionUuid: dto.sessionUuid,
    cashRegisterUuid: dto.cashRegisterUuid,
    cashRegisterName: dto.cashRegisterName,
    cashRegisterLocation: dto.cashRegisterLocation,
    openedAt: dto.openedAt,
    openedBy: dto.openedBy,
    openingAmount: dto.openingAmount,
    totalCashPayments: dto.totalCashPayments,
    expectedCashAmount: dto.expectedCashAmount,
    totalCashMovements: dto.totalCashMovements,
    paymentMethodSummaries: dto.paymentMethodSummaries.map(mapToPaymentMethodSummary)
  };
}

File: gastromanager-backend/src/main/java/com/kaiho/gastromanager/application/cashregistersession/mapper/CashRegisterSessionMapper.java
package com.kaiho.gastromanager.application.cashregistersession.mapper;

import com.kaiho.gastromanager.application.cashregister.dto.response.CashRegisterCurrentSessionResponseDto;
import com.kaiho.gastromanager.application.cashregistersession.dto.request.CashRegisterSessionCloseRequestDto;
import com.kaiho.gastromanager.application.cashregistersession.dto.request.CashRegisterSessionOpenRequestDto;
import com.kaiho.gastromanager.application.cashregistersession.dto.response.CashRegisterSessionResponseDto;
import com.kaiho.gastromanager.application.cashregistersession.dto.response.CashRegisterSessionSummaryResponseDto;
import com.kaiho.gastromanager.application.cashregistersession.dto.response.PaymentMethodSummaryDto;
import com.kaiho.gastromanager.domain.cashregister.model.CashRegister;
import com.kaiho.gastromanager.domain.cashregistersession.model.CashRegisterSession;
import com.kaiho.gastromanager.domain.restaurant.model.Restaurant;
import org.springframework.stereotype.Component;

import java.util.List;

import static com.kaiho.gastromanager.infrastructure.config.context.RestaurantContext.getCurrentRestaurant;


@Component
public class CashRegisterSessionMapper {

    public CashRegisterSession toDomain(CashRegisterSessionOpenRequestDto requestDto) {
        if (requestDto == null) {
            return null;
        }

        CashRegister cashRegister = CashRegister.builder()
                                                .uuid(requestDto.cashRegisterUuid())
                                                .build();

        Restaurant restaurant = Restaurant.builder()
                                          .uuid(getCurrentRestaurant())
                                          .build();

        return CashRegisterSession.builder()
                                  .cashRegister(cashRegister)
                                  .restaurant(restaurant)
                                  .openingAmount(requestDto.openingAmount())
                                  .notes(requestDto.notes())
                                  .build();
    }

    public CashRegisterSession toDomain(CashRegisterSessionCloseRequestDto requestDto) {
        if (requestDto == null) {
            return null;
        }

        return CashRegisterSession.builder()

                                  .closingAmount(requestDto.closingAmount())
                                  .notes(requestDto.notes())
                                  .build();
    }

    public CashRegisterSessionResponseDto toResponse(CashRegisterSession session) {
        if (session == null) {
            return null;
        }

        return CashRegisterSessionResponseDto.builder()
                                             .uuid(session.getUuid())
                                             .cashRegisterUuid(session.getCashRegister().getUuid())
                                             .cashRegisterName(session.getCashRegister().getName())
                                             .cashRegisterLocation(session.getCashRegister().getLocation())
                                             .openingTime(session.getOpeningTime())
                                             .closingTime(session.getClosingTime())
                                             .openingAmount(session.getOpeningAmount())
                                             .closingAmount(session.getClosingAmount())
                                             .status(session.getStatus())
                                             .createdBy(session.getCreatedBy())
                                             .createdDate(session.getCreatedDate())
                                             .build();
    }

    public CashRegisterCurrentSessionResponseDto toCurrentSessionResponse(CashRegisterSession session) {
        if (session == null) {
            return null;
        }

        return CashRegisterCurrentSessionResponseDto.builder()
                                                    .uuid(session.getUuid())
                                                    .openedAt(session.getOpeningTime())
                                                    .openedBy(session.getCreatedBy())
                                                    .openingAmount(session.getOpeningAmount())
                                                    .build();
    }

    public CashRegisterSessionSummaryResponseDto toSummaryResponse(CashRegisterSession session) {
        if (session == null) {
            return null;
        }

        // Mapear los resúmenes de métodos de pago
        List<PaymentMethodSummaryDto> paymentMethodDtos = session.getPaymentMethodSummaries() != null ?
                session.getPaymentMethodSummaries().stream()
                       .map(paymentSummary -> PaymentMethodSummaryDto.builder()
                                                                     .paymentMethodName(paymentSummary.paymentMethodName())
                                                                     .paymentMethodDescription(paymentSummary.paymentMethodDescription())
                                                                     .totalAmount(paymentSummary.totalAmount())
                                                                     .transactionCount(paymentSummary.transactionCount())
                                                                     .build())
                       .toList() : List.of();

        return CashRegisterSessionSummaryResponseDto.builder()
                                                    .sessionUuid(session.getUuid())
                                                    .cashRegisterUuid(session.getCashRegister().getUuid())
                                                    .cashRegisterName(session.getCashRegister().getName())
                                                    .cashRegisterLocation(session.getCashRegister().getLocation())
                                                    .openedAt(session.getOpeningTime())
                                                    .openedBy(session.getCreatedBy())
                                                    .openingAmount(session.getOpeningAmount())
                                                    .totalCashPayments(session.getTotalCashPayments())
                                                    .expectedCashAmount(session.getExpectedAmount())
                                                    .totalCashMovements(session.getTotalCashMovements())
                                                    .paymentMethodSummaries(paymentMethodDtos)
                                                    .build();
    }
}
